src/block/WaterCauldron.php

```diff
@@ -137,8 +137,14 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 					ItemTypeIds::LEATHER_PANTS,
 					ItemTypeIds::LEATHER_BOOTS => true,
 					default => false
-				} && $item->getCustomColor()?->toRGBA() !== $this->customWaterColor->toRGBA()){
-					$item->setCustomColor($this->customWaterColor);
+				} && $item->getCustomColor()?->toRGBA() !== $this->customWaterColor->toRGBA() && $player !== null){
+					$newItem = (clone $item)->setCustomColor($this->customWaterColor);
+					$item->pop();
+					if($item->getCount() === 0){
+						$player->getInventory()->setItemInHand($newItem);
+					}else{ //if the stack size was bigger than 1 (usually won't happen, but might be caused by plugins
+						$player->getInventory()->addItem($newItem);
+					}
```

Comment:
Use `$returnedItems` for this.

Reply:
The use of returnedItems will result in non-vanilla behavior, because in Vanilla the item is set to the player's current slot